### PR TITLE
Feature/loki log forwarding

### PR DESCRIPTION
**Pull request by computate**
I managed to get Log Forwarding working with the OpenShift Logging Operator, Loki Operator, Grafana Operator, and required Elasticsearch Operator. Loki doesn't support using the service-ca cert for TLS authentication to the S3 storage proved by the Data Foundation Operator, and other important configuration changes, so I setup custom Loki Deployments and StatefulSets that can be customized with the service-ca cert and other Loki configuration settings. I made sure the corresponding Loki operator provided Deployments and StatefulSets are not started. 

This isn't a great solution to have the extra Loki Deployments and StatefulSets, so I want to still work with Red Hat Support and the Loki team to see what they can help us with there. 

Also, grafana here is not yet integrated with the single sign on of OpenShift. You can find the Grafana username and password to login here: 
https://logging-grafana.apps.nerc-ocp-infra.rc.fas.harvard.edu

Look at the grafana-admin-credentials for the username and password with the command below. You will also need to `base64 -d` decode the GF_SECURITY_ADMIN_USER and GF_SECURITY_ADMIN_PASSWORD. 
```bash
oc --as system:admin -n grafana get secret/grafana-admin-credentials -o yaml
```

Please let me know what you think so far. 

I have created an feature request issue on the loki side to ask them to support the ca_file and other configuration settings in the LokiOperator here: 

https://github.com/grafana/loki/issues/6932

Documentation for this task can be found in many links here: 

- https://loki-operator.dev/docs/forwarding_logs_to_gateway.md/
- https://grafana.com/docs/loki/latest/configuration/
- https://access.redhat.com/documentation/en-us/openshift_container_platform/4.10/html/logging/cluster-logging-external#cluster-logging-collector-log-forward-loki_cluster-logging-external
- https://access.redhat.com/documentation/en-us/openshift_container_platform/4.10/html-single/logging/index#cluster-logging-exported-fields-kubernetes_cluster-logging-exported-fields
- https://docs.openshift.com/container-platform/4.10/logging/cluster-logging-exported-fields.html#cluster-logging-exported-fields-kubernetes_cluster-logging-exported-fields
- https://docs.openshift.com/container-platform/4.10/logging/cluster-logging-deploying.html
- https://grafana.com/docs/loki/latest/api/#push-log-entries-to-loki

### CODE REVIEW

**Review by naved001 — COMMENTED** [review]
Thanks @computate! This is a lot of work.

Here are my initial thoughts:

* There should be an argocd app for Loki and Grafana. And that app should be deploying all the namespaced resources for loki/grafana like deployments, routes, externalsecrets etc. The `cluster-scope` app should only be used for deploying clusterwide things (clusterrolebindings, creating new namespaces, operator subscriptions etc). 
* It would be nice if we could organize this PR into multiple commits to make it easier to review. I will not block the PR on that, that's merely a suggestion.
* Could you link to some documentation that you referred to while creating these resources? That way I can better understand whats going on in this PR. Ideally the link to relevant documentation should be a part of the commit message.

I will leave a more thorough review soon-ish.

**Comment by computate** [comment]
Thanks for the suggestions @naved001 @joachimweyl , I can take a look at implementing your suggestions after I get back from vacation on Monday. Meanwhile, I have added all the useful documentation links in the top description of the PR.

File: cluster-scope/base/core/namespaces/openshift-logging/namespace.yaml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: openshift-logging
+  annotations:
+    openshift.io/node-selector: ""
+  labels:
+    openshift.io/cluster-monitoring: 'true'
+spec: {}
```
Comment (naved001) [inline cluster-scope/base/core/namespaces/openshift-logging/namespace.yaml:9]: why is `spec` specified here? I didn't see it in the redhat document that describe this namespace.

Comment (naved001) [reply cluster-scope/base/core/namespaces/openshift-logging/namespace.yaml:9]: same thing with the other namespaces. Just curious what does setting `spec` to `{}` do for a namepsace.

File: cluster-scope/base/operators.coreos.com/subscriptions/grafana-operator/subscription.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: loki-operator
```
Comment (naved001) [inline cluster-scope/base/operators.coreos.com/subscriptions/grafana-operator/subscription.yaml:4]: this should be `grafana-operator`

File: cluster-scope/base/operators.coreos.com/subscriptions/cluster-logging/subscription.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: loki-operator
```
Comment (naved001) [inline cluster-scope/base/operators.coreos.com/subscriptions/cluster-logging/subscription.yaml:4]: this should be `cluster-logging`

File: cluster-scope/base/external-secrets.io/externalsecrets/lokistack-config-custom/externalsecret.yaml [hunk 4]
```diff
@@ -0,0 +1,191 @@
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+namespace: openshift-operators-redhat
```
Comment (naved001) [inline cluster-scope/base/external-secrets.io/externalsecrets/lokistack-config-custom/externalsecret.yaml:3]: `namespace` should only be under `metadata`. Specifying it here makes this invalid and this manifest won't be deployed.

File: cluster-scope/base/external-secrets.io/externalsecrets/openshift-logging-lokistack-gateway-bearer-token/externalsecret.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+namespace: openshift-logging
```
Comment (naved001) [inline cluster-scope/base/external-secrets.io/externalsecrets/openshift-logging-lokistack-gateway-bearer-token/externalsecret.yaml:3]: same; `namespace` should be removed from here.

File: cluster-scope/overlays/nerc-ocp-infra/externalsecrets/openshift-logging-lokistack-gateway-bearer-token_patch.yaml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+namespace: openshift-logging
```
Comment (naved001) [inline cluster-scope/overlays/nerc-ocp-infra/externalsecrets/openshift-logging-lokistack-gateway-bearer-token_patch.yaml:3]: `namespace` should be removed from here.

File: cluster-scope/overlays/nerc-ocp-infra/externalsecrets/loki-config-custom_patch.yaml [hunk 7]
```diff
@@ -0,0 +1,24 @@
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+namespace: openshift-operators-redhat
```
Comment (naved001) [inline cluster-scope/overlays/nerc-ocp-infra/externalsecrets/loki-config-custom_patch.yaml:3]: `namespace` should be removed from here.

**Review by naved001 — CHANGES_REQUESTED** [review]

**Comment by computate** [comment]
I'm going to close this PR and simplify these updates into smaller PRs based on Naved's great suggestion for argocd apps.

**Comment by joachimweyl** [comment]
@computate please link smaller PR's here.

**Comment by computate** [comment]
@joachimweyl Here are the new smaller prs: 
https://github.com/OCP-on-NERC/nerc-ocp-config/pull/69
https://github.com/OCP-on-NERC/nerc-ocp-config/pull/70
https://github.com/OCP-on-NERC/nerc-ocp-config/pull/71
https://github.com/OCP-on-NERC/nerc-ocp-apps/pull/9

**Pull request closed without merging**